### PR TITLE
Make `serve` generic over the listener and IO types

### PR DESCRIPTION
**Pull request by davidpdrsn**
Based on https://github.com/tokio-rs/axum/pull/2478

Fixes https://github.com/tokio-rs/axum/issues/2474

It's actually not as bad as I had feared so I think we can do this for 0.8.

### CODE REVIEW

File: axum/src/serve.rs
Comment (davidpdrsn) [inline axum/src/serve.rs:41]: We could also use tokio-util's [`Listener`](https://docs.rs/tokio-util/latest/tokio_util/net/trait.Listener.html) trait however I don't think tokio-util should be in our public API. By writing our own we can also use `async fn accept` rather than `fn poll_accept`.

File: examples/unix-domain-socket/src/main.rs
Comment (jplatte) [inline examples/unix-domain-socket/src/main.rs:95]: Is there some reason something like this can't be part of axum? I think we could implement at least `impl Connected<unix::SocketAddr> for unix::SocketAddr`, `impl Connected<IncomingStream<'_, UnixListener>> for unix::SocketAddr` and `impl Connected<IncomingStream<'_, UnixListener>> for UCreds`, right?

Comment (davidpdrsn) [reply examples/unix-domain-socket/src/main.rs:95]: You cannot do `impl Connected<unix::SocketAddr> for unix::SocketAddr` because `unix::SocketAddr` is not `Clone`.

But we could have some `UdsConnectInfo` in axum that uses `Arc<unix::SocketAddr>` (similarly to [what tonic has](https://docs.rs/tonic/latest/tonic/transport/server/struct.UdsConnectInfo.html)) but I've never actually used unix streams myself so I'm not sure what info that type should contain. Do you have an opinion about that?

Comment (jplatte) [reply examples/unix-domain-socket/src/main.rs:95]: Hm, no. I've used UDS w/ axum once, but didn't care about connection info then.

Comment (davidpdrsn) [reply examples/unix-domain-socket/src/main.rs:95]: In that case I think we should just leave it as is. We can always add it if someone requests it and has a use case.

**Review by jplatte — APPROVED** [review]

**Comment by davidpdrsn** [comment]
> Is any help needed to bring this change over the finish line?

Nope. Its a breaking change and don't wanna start merging breaking PRs just yet. It is done otherwise.

**Comment by lojewalo** [comment]
> Is any help needed to bring this change over the finish line?

I'm not involved with the project, but seeing as this is a breaking change, it's slotted for 0.8 as far as I can tell, so not sure how soon it will be until this is merged.

As a workaround for myself, I cloned the project and merged this PR specifically into the latest version. If it's viable to use for your situation, you can patch it in via the below snippet. I'm using this in prod with no issues, but YMMV obviously.

```toml
# in your Cargo.toml
[patch.crates-io]
axum = { git = "https://github.com/lojewalo/axum.git", branch = "uds-v0.7.3" }
```

**Comment by kamulos** [comment]
I came here via the issue #2494. If this makes it easier to accept tls connections, could you also provide an example how to do it without the `axum-server` crate?

I am at the moment a bit frustrated about the slow maintenance of the `axum-server` crate and would prefer to avoid depending on it.

**Comment by elichai** [comment]
Any updates on where this stands with the maintainers? do you want to experiment with different designs or can this be merged as-is? 
This also solves for me a case of listening over VSOCK with something a little bit different than the regular TCP syscall-flow

**Comment by jplatte** [comment]
This is a PR by the primary maintainer, approved by the co-maintainer. It's going to land, the question is just when. Which I unfortunately can't give you any details about, you're just going to have to be patient (or use this branch via [`[patch]`](https://doc.rust-lang.org/cargo/reference/overriding-dependencies.html#the-patch-section)).

**Comment by xtexChooser** [comment]
This PR is added to 0.8 milestone. Will it land when 0.8 publishes?

**Comment by jplatte** [comment]
Yes.

**Comment by jplatte** [comment]
Closing in favor of #2941.

**Pull request closed without merging**
